src/coreclr/nativeaot/Runtime/thread.inl

```diff
@@ -42,17 +42,40 @@ inline PTR_VOID Thread::GetThreadStressLog() const
     return m_pThreadStressLog;
 }
 
-inline void Thread::EnterCantAllocRegion()
+inline void Thread::EnterForbidBlockingRegion()
 {
-    m_cantAlloc++;
+#ifdef _DEBUG
+    m_forbidBlocking++;
+#endif
 }
 
-inline void Thread::LeaveCantAllocRegion()
+inline void Thread::LeaveForbidBlockingRegion()
 {
-    m_cantAlloc--;
+#ifdef _DEBUG
+    m_forbidBlocking--;
+#endif
 }
 
-inline bool Thread::IsInCantAllocStressLogRegion()
+inline bool Thread::IsInForbidBlockingRegion()
 {
-    return m_cantAlloc != 0;
+#ifdef _DEBUG
+    return m_forbidBlocking != 0;
+#else
+    return false;
+#endif
 }
+
+struct ForbidBlockingHolder
```

Comment:
I would stay away from introducing C++ into NativeAOT runtime. In the ideal case, the NativeAOT runtime would not depend on C++ runtime at all. There are not that many places that would have to be cleaned up to make it happen. https://github.com/dotnet/corert/issues/3564 has some old discussion on this.

Reply:
> I would stay away from introducing C++ into NativeAOT runtime. 

I meant introducing constructs that pull in C++ runtime.

Using C++ as "C with classes" is fine.

Reply:
I've seen holders are used in other places, so assumed it is in-style. This particular case can easily do without a holder though.

Reply:
Yes, avoiding c++ runtime and using C-with-classes is reasonable. Some cases are obvious. We do not want `std::this::kind::of::code` or exception handling.

Holders are confusing though. In theory destructors are exception-aware constructs, but I am not sure if they necessarily involve runtime support. It does not look so when I look at emitted assembly. 
We often use destructors just to not write the same code on every exit from a scope.

Reply:
Methods with on-stack destructors reference C++ exception handling routine in unwind data.

Reply:
I see. I suspected that. Anyways, I've removed the holder.